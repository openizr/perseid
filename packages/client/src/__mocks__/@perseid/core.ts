/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/core` mock.
 */

export class HttpClient {
  protected mock = vi.fn();

  protected async request(): Promise<void> {
    await this.mock();
  }
}
export class Model { protected mocked = vi.fn(); }
export class Logger { protected mocked = vi.fn(); }
export class Id { public toString = vi.fn(() => '000000000000000000000001'); }
export const toSnakeCase = vi.fn((variable: string) => `TO_SNAKE_CASE_${variable}`);
export const deepMerge = vi.fn((_: unknown, variable: unknown): unknown => variable);
export const deepCopy = vi.fn((variable: unknown): unknown => (typeof variable === 'object' ? { ...variable } : variable));
export const isPlainObject = vi.fn((variable: unknown): boolean => typeof variable === 'object' && !Array.isArray(variable) && !(variable instanceof Id));
