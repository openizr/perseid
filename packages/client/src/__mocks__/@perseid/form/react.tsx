/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/form/react` mock.
 */

export default function Form(props: unknown): JSX.Element {
  return (
    <div id="react-form">
      <span>{JSON.stringify(props)}</span>
    </div>
  );
}
