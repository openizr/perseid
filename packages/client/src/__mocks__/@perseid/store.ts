/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/store` mock.
 */

export default class Store {
  public register = vi.fn();

  public mutate = vi.fn();

  public subscribe = vi.fn();

  public combine = vi.fn();

  public dispatch = vi.fn();
}
