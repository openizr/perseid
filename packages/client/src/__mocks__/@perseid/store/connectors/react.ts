/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/store/connectors/react` mock.
 */

export default vi.fn(() => vi.fn((module: string) => {
  if (module === 'router') {
    return (process.env.NOT_FOUND === 'true')
      ? { route: null }
      : { route: process.env.PAGE };
  }
  return (process.env.ERROR === 'true')
    ? new Error('test')
    : null;
}));
