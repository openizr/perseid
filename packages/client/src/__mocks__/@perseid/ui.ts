/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * `@perseid/ui` mock.
 */

export const test = true;
export const generateRandomId = vi.fn(() => '18972182');
