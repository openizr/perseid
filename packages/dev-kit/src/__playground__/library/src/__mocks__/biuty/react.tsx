import React from 'react';

export function UIP(): JSX.Element {
  return <p>TEST</p>;
}

export function UITitle(): JSX.Element {
  return <p>TEST</p>;
}
