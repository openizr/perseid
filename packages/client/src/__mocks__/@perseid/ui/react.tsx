/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';

/**
 * `@perseid/ui/react` mock.
 */

interface ComponentProps {
  [key: string]: unknown;
  onClick?: (event: Event) => void;
  onBlur?: (value: string) => void;
  onFocus?: (value: string) => void;
  onChange?: (newValue: string) => void;
  onKeyDown?: (value: string, event: KeyboardEvent) => void;
  columns: { component: React.ReactNode; path: string; }[];
}

function Component(type: string): (props: ComponentProps) => JSX.Element {
  return function RealComponent({
    onBlur,
    onClick,
    onFocus,
    onChange,
    onKeyDown,
    ...props
  }) {
    React.useEffect(() => {
      // Covers `onChange` handler.
      if (onChange !== undefined) {
        onChange((props.value === '2023-01-01T00:00:00.000Z')
          ? '2023-01-02T00:00:00.000Z'
          : process.env.VALUE ?? 'test');
      }
      // Covers `onBlur` handler.
      if (onBlur !== undefined) {
        onBlur('test');
      }
      // Covers `onKeyDown` handler.
      if (onKeyDown !== undefined) {
        onKeyDown('test', { key: 'ArrowDown' } as KeyboardEvent);
      }
      // Covers `onFocus` handler.
      if (onFocus !== undefined) {
        onFocus('test');
      }
      // Covers `onClick` handler.
      if (onClick !== undefined) {
        onClick({ preventDefault: vi.fn() } as unknown as Event);
      }
    }, []);
    if (type === 'ui-options') {
      return (
        <div id={type}>
          {JSON.stringify(props)}
          <span>
            <span />
            <span>
              <span />
            </span>
          </span>
        </div>
      );
    }
    if (type === 'ui-textfield') {
      return (
        <div id={type} onFocus={onFocus as unknown as React.FocusEventHandler}>
          {JSON.stringify(props)}
          <span className="ui-textfield__wrapper__field" />
        </div>
      );
    }
    return <div id={type}>{JSON.stringify(props)}</div>;
  };
}

export const UIP = Component('ui-p');
export const UILink = Component('ui-link');
export const UIImage = Component('ui-image');
export const UITitle = Component('ui-title');
export const UIButton = Component('ui-button');
export const UIOptions = Component('ui-options');
export const UITextarea = Component('ui-textarea');
export const UITextfield = Component('ui-textfield');
export const UIFilePicker = Component('ui-file-picker');
export const generateRandomId = vi.fn(() => '18972182');
export const buildClass = vi.fn((...values: string[]): string => values.join(' '));
export const markdown = vi.fn((label: string, lightMode: boolean) => `MARKDOWN FOR ${label}, ${lightMode}`);
