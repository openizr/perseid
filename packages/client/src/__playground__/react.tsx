/* c8 ignore start */

import {
  Model,
  Store,
  Logger,
  ApiClient,
  FormBuilder,
} from 'scripts/core/index';
import {
  Grid,
  Router,
} from 'scripts/react/index';
import {
  type Ids,
  type Authors,
  type Version,
  type Deletion,
  type Timestamps,
  type DefaultDataModel, I18n,
} from '@perseid/core';
import * as React from 'react';
import '__playground__/style.scss';
import labels from '__playground__/labels';
import { createRoot, type Root } from 'react-dom/client';
import { type StoreSettings } from 'scripts/core/services/Store';
import type { UseSubscription } from '@perseid/store/connectors/react';

let app: Root;

interface DataModel extends DefaultDataModel {
  tests: Ids & Deletion & Timestamps & Authors & Version & {
    elements: null | {
      child: string;
    }[];
  };
}

function main(): void {
  const logger = new Logger({ logLevel: 'debug' });
  const i18n = new I18n(logger, labels);
  const pages: StoreSettings<DataModel>['pages'] = {
    auth: {
      signIn: { route: '/sign-in' },
      signUp: { route: '/sign-up' },
      updateUser: { route: '/users/me' },
      verifyEmail: { route: '/verify-email' },
      resetPassword: { route: '/reset-password' },
    },
    resources: {
      users: {
        list: {
          route: '/users',
          pageProps: {
            searchFields: ['email'],
            fields: [
              '_createdAt',
              'email',
              'roles._createdBy._updatedBy',
              '_devices',
              'password',
              '_apiKeys',
              '_verifiedAt',
            ],
          },
        },
        view: {
          route: '/users/:id',
          pageProps: {
            fields: [
              '_createdAt',
              'roles._createdBy.email',
              '_devices',
              'password',
              '_apiKeys',
              '_verifiedAt',
            ],
          },
        },
        create: {
          route: '/users/create',
          pageProps: {
            fields: ['roles.name'],
          },
        },
        update: {
          route: '/users/:id/edit',
          pageProps: {
            fields: ['roles.name'],
          },
        },
      },
      roles: {
        list: {
          route: '/roles',
          pageProps: {
            fields: ['name'],
            searchFields: ['name'],
          },
        },
        view: {
          route: '/roles/:id',
          pageProps: {
            fields: ['_createdAt', '_createdBy', 'name', 'permissions'],
          },
        },
        update: {
          route: '/roles/:id/edit',
          pageProps: {},
        },
        create: {
          route: '/roles/create',
          pageProps: {},
        },
      },
      tests: {
        list: {
          route: '/tests',
          pageProps: {
            fields: [
              '_createdAt',
              '_createdBy.email',
              'elements',
            ],
          },
        },
        view: {
          route: '/tests/:id',
          pageProps: {
            fields: [
              '_createdAt',
              '_createdBy.email',
              'elements',
            ],
          },
        },
        create: {
          route: '/tests/create',
          pageProps: {
            fields: [],
          },
        },
        update: {
          route: '/tests/:id/edit',
          pageProps: {
            fields: [],
          },
        },
      },
    },
  };
  const model = new Model<DataModel>();
  const apiClient = new ApiClient<DataModel>(model, logger, {
    endpoints: {
      auth: {
        viewMe: { route: '/auth/me' },
        signUp: { route: '/auth/sign-up' },
        signIn: { route: '/auth/sign-in' },
        signOut: { route: '/auth/sign-out' },
        verifyEmail: { route: '/auth/verify-email' },
        refreshToken: { route: '/auth/refresh-token' },
        resetPassword: { route: '/auth/reset-password' },
        requestPasswordReset: { route: '/auth/reset-password' },
        requestEmailVerification: { route: '/auth/verify-email' },
      },
      resources: {
        users: {
          list: { route: '/users' },
          create: { route: '/users' },
          view: { route: '/users/:id' },
          update: { route: '/users/:id' },
          search: { route: '/users/search' },
        },
        roles: {
          list: { route: '/roles' },
          create: { route: '/roles' },
          view: { route: '/roles/:id' },
          update: { route: '/roles/:id' },
          search: { route: '/roles/search' },
        },
        tests: {
          list: { route: '/tests' },
          create: { route: '/tests' },
          view: { route: '/tests/:id' },
          update: { route: '/tests/:id' },
          delete: { route: '/tests/:id' },
          search: { route: '/tests/search' },
        },
      },
    },
    mockedResponses: {},
    connectTimeout: 3000,
    baseUrl: 'http://localhost:5070/perseid',
  });
  const formBuilder = new FormBuilder<DataModel>(model, logger);
  const store = new Store<DataModel>(model, logger, apiClient, formBuilder, {
    pages,
    fallbackPageRoute: '/',
  });
  store.createRoute('/', {
    component: 'Home',
    visibility: 'PRIVATE',
  });
  store.createRoute('/ok', {
    component: 'Home',
    visibility: 'PUBLIC',
  });
  store.createRoute('/ko', {
    component: 'Home',
    visibility: 'PUBLIC_ONLY',
  });
  store.createRoutes();
  const container = document.querySelector('#root') as unknown as HTMLElement;
  app = createRoot(container);
  app.render(
    <React.StrictMode>
      <Router<DataModel>
        services={{
          i18n,
          model,
          store: store as Store<DataModel> & { useSubscription: UseSubscription; },
          apiClient,
        }}
        pages={{
          Home: async () => Promise.resolve({ default: () => <div /> }),
        }}
        components={{}}
      />
      <Grid columns={{ mobile: 4, tablet: 8, desktop: 12 }} />
    </React.StrictMode>,
  );
}

// Ensures DOM is fully loaded before running app's main logic.
// Loading hasn't finished yet...
if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', main);
  // `DOMContentLoaded` has already fired...
} else {
  main();
}

// Ensures subscriptions to Store are correctly cleared when page is left, to prevent "ghost"
// processing, by manually unmounting React components tree.
window.addEventListener('beforeunload', () => {
  app.unmount();
});
