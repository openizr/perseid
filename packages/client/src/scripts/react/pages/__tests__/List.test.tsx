/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @vitest-environment jsdom
 */

import { render } from '@testing-library/react';
import { type DefaultDataModel } from '@perseid/core';
import List from 'scripts/react/pages/List';

type Services = CommonProps<DefaultDataModel>['services'];

describe('react/pages/List', () => {
  vi.mock('@perseid/core');
  vi.mock('@perseid/ui/react');
  vi.mock('scripts/react/components/Table');
  vi.mock('scripts/react/components/Loader');
  vi.mock('scripts/react/components/FieldValue');
  vi.mock('scripts/react/components/FieldLabel');
  vi.mock('scripts/react/components/PageLayout');
  vi.mock('scripts/react/components/Pagination');

  const components = {};
  const goToPage = vi.fn();
  const listOrSearch = vi.fn();
  const createServices = (
    pageData: unknown,
    registryData: unknown,
    viewRoute: string | null,
  ): Services => ({
    apiClient: {},
    model: { get: vi.fn((path: string) => ({ schema: { isIndexed: path.includes('sortable') } })) },
    i18n: { t: vi.fn((label: string) => label) },
    store: {
      goToPage,
      listOrSearch,
      navigate: vi.fn(),
      getRoute: vi.fn(() => viewRoute),
      useSubscription: vi.fn((subscription) => (
        (subscription === 'registry') ? registryData : pageData
      )),
    },
  }) as unknown as Services;

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly - loading page', () => {
    const { container } = render(
      <List
        resource="users"
        components={components}
        services={createServices({ results: null }, null, null)}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - loaded page, results, view route exists', () => {
    const searchBody = { query: { on: ['searchField'], text: 'test' }, filters: null };
    const { container } = render(
      <List
        resource="users"
        components={components}
        services={createServices({
          page: 1,
          limit: 10,
          total: 20,
          search: null,
          sorting: null,
          loading: false,
          searchFields: ['searchField'],
          fields: ['sortableField', 'nonSortableField'],
          results: ['000000000000000000000011', '123456789012345678901235'],
        }, {
          users: {
            '000000000000000000000011': { sortableField: 'Value 1', nonSortableField: 'Value 2' },
          },
        }, '/view/:id')}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(listOrSearch).toHaveBeenCalledTimes(2);
    expect(listOrSearch).toHaveBeenCalledWith('users', null, {
      page: 1,
      limit: 10,
      total: 20,
      search: null,
      loading: false,
      sorting: { field1: 1 },
      searchFields: ['searchField'],
      fields: ['sortableField', 'nonSortableField'],
      results: ['000000000000000000000011', '123456789012345678901235'],
    });
    expect(listOrSearch).toHaveBeenCalledWith('users', searchBody, {
      page: 1,
      limit: 10,
      total: 20,
      search: null,
      sorting: null,
      loading: false,
      searchFields: ['searchField'],
      fields: ['sortableField', 'nonSortableField'],
      results: ['000000000000000000000011', '123456789012345678901235'],
    });
    expect(goToPage).toHaveBeenCalledTimes(1);
    expect(goToPage).toHaveBeenCalledWith({
      page: 1,
      limit: 10,
      total: 20,
      search: null,
      sorting: null,
      loading: false,
      searchFields: ['searchField'],
      fields: ['sortableField', 'nonSortableField'],
      results: ['000000000000000000000011', '123456789012345678901235'],
    });
  });

  test('renders correctly - loaded page, results, view route does not exist', () => {
    const { container } = render(
      <List
        resource="users"
        components={components}
        services={createServices({
          page: 1,
          limit: 10,
          total: 20,
          search: null,
          sorting: null,
          loading: false,
          searchFields: ['searchField'],
          fields: ['sortableField', 'nonSortableField'],
          results: ['000000000000000000000011', '123456789012345678901235'],
        }, {
          users: {
            '000000000000000000000011': { sortableField: 'Value 1', nonSortableField: 'Value 2' },
          },
        }, null)}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - loaded page, no results', () => {
    const { container } = render(
      <List
        resource="users"
        components={components}
        services={createServices({
          results: [],
          fields: ['sortableField', 'nonSortableField'],
          searchFields: ['searchField'],
          sorting: null,
          page: 1,
          limit: 10,
          total: 0,
          loading: false,
          search: { query: { text: null } },
        }, {}, null)}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
