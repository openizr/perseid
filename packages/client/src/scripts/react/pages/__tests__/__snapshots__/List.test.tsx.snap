// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react/pages/List > renders correctly - loaded page, no results 1`] = `
<main
  class="list-page users"
>
  <div
    id="page-layout"
  >
    <span>
      {"page":"LIST","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
    </span>
    <div>
      <div
        id="ui-title"
      >
        {"label":"PAGES.TO_SNAKE_CASE_users.LIST.TITLE"}
      </div>
      <div
        class="list-page__search"
      >
        <div
          id="ui-textfield"
        >
          {"autofocus":true,"name":"search","maxlength":50,"id":"search-bar","autocomplete":"off","debounceTimeout":250,"value":"","placeholder":"PAGES.TO_SNAKE_CASE_users.LIST.SEARCH_PLACEHOLDER"}
          <span
            class="ui-textfield__wrapper__field"
          />
        </div>
      </div>
      <div
        class="list-page__content"
      >
        <div
          id="table"
        >
          <span>
            {"labels":{"loading":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.LOADING","actions":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.ACTIONS","noResult":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.NO_RESULT"},"sorting":null}
          </span>
          <div>
            <div>
              <span>
                {"path":"sortableField","isSortable":true}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"sortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
            <div>
              <span>
                {"path":"nonSortableField","isSortable":false}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"nonSortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
          </div>
          <div />
        </div>
        <div
          id="pagination"
        >
          {"services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"total":0,"components":{},"resource":"users","currentPage":1,"itemsPerPage":10}
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`react/pages/List > renders correctly - loaded page, results, view route does not exist 1`] = `
<main
  class="list-page users"
>
  <div
    id="page-layout"
  >
    <span>
      {"page":"LIST","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
    </span>
    <div>
      <div
        id="ui-title"
      >
        {"label":"PAGES.TO_SNAKE_CASE_users.LIST.TITLE"}
      </div>
      <div
        class="list-page__search"
      >
        <div
          id="ui-textfield"
        >
          {"autofocus":true,"name":"search","maxlength":50,"id":"search-bar","autocomplete":"off","debounceTimeout":250,"value":"","placeholder":"PAGES.TO_SNAKE_CASE_users.LIST.SEARCH_PLACEHOLDER"}
          <span
            class="ui-textfield__wrapper__field"
          />
        </div>
      </div>
      <div
        class="list-page__content"
      >
        <div
          id="table"
        >
          <span>
            {"labels":{"loading":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.LOADING","actions":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.ACTIONS","noResult":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.NO_RESULT"},"sorting":null}
          </span>
          <div>
            <div>
              <span>
                {"path":"sortableField","isSortable":true}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"sortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
            <div>
              <span>
                {"path":"nonSortableField","isSortable":false}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"nonSortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
          </div>
          <div>
            <div>
              <div
                id="field-value"
              >
                {"id":"000000000000000000000011","page":"LIST","field":"sortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"registry":{"users":{"000000000000000000000011":{"sortableField":"Value 1","nonSortableField":"Value 2"}}},"resource":"users","components":{},"loading":false}
              </div>
              <div
                id="field-value"
              >
                {"id":"000000000000000000000011","page":"LIST","field":"nonSortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"registry":{"users":{"000000000000000000000011":{"sortableField":"Value 1","nonSortableField":"Value 2"}}},"resource":"users","components":{},"loading":false}
              </div>
            </div>
          </div>
        </div>
        <div
          id="pagination"
        >
          {"services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"total":20,"components":{},"resource":"users","currentPage":1,"itemsPerPage":10}
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`react/pages/List > renders correctly - loaded page, results, view route exists 1`] = `
<main
  class="list-page users"
>
  <div
    id="page-layout"
  >
    <span>
      {"page":"LIST","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
    </span>
    <div>
      <div
        id="ui-title"
      >
        {"label":"PAGES.TO_SNAKE_CASE_users.LIST.TITLE"}
      </div>
      <div
        class="list-page__search"
      >
        <div
          id="ui-textfield"
        >
          {"autofocus":true,"name":"search","maxlength":50,"id":"search-bar","autocomplete":"off","debounceTimeout":250,"value":"","placeholder":"PAGES.TO_SNAKE_CASE_users.LIST.SEARCH_PLACEHOLDER"}
          <span
            class="ui-textfield__wrapper__field"
          />
        </div>
      </div>
      <div
        class="list-page__content"
      >
        <div
          id="table"
        >
          <span>
            {"labels":{"loading":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.LOADING","actions":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.ACTIONS","noResult":"PAGES.TO_SNAKE_CASE_users.LIST.TABLE.NO_RESULT"},"sorting":null}
          </span>
          <div>
            <div>
              <span>
                {"path":"sortableField","isSortable":true}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"sortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
            <div>
              <span>
                {"path":"nonSortableField","isSortable":false}
              </span>
              <div
                id="field-label"
              >
                {"page":"LIST","field":"nonSortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"resource":"users","components":{}}
              </div>
            </div>
          </div>
          <div>
            <div>
              <div
                id="field-value"
              >
                {"id":"000000000000000000000011","page":"LIST","field":"sortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"registry":{"users":{"000000000000000000000011":{"sortableField":"Value 1","nonSortableField":"Value 2"}}},"resource":"users","components":{},"loading":false}
              </div>
              <div
                id="field-value"
              >
                {"id":"000000000000000000000011","page":"LIST","field":"nonSortableField","services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"registry":{"users":{"000000000000000000000011":{"sortableField":"Value 1","nonSortableField":"Value 2"}}},"resource":"users","components":{},"loading":false}
              </div>
            </div>
          </div>
        </div>
        <div
          id="pagination"
        >
          {"services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"total":20,"components":{},"resource":"users","currentPage":1,"itemsPerPage":10}
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`react/pages/List > renders correctly - loading page 1`] = `
<div
  id="loader"
>
  {"services":{"apiClient":{},"model":{},"i18n":{},"store":{}},"components":{}}
</div>
`;
